keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -907,22 +917,26 @@ Problem examples:
 | CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
 | Validation of an aggregated API server type            | Main API server does not have type definitions |
 
-Due to these complications, we have decided to evalute CEL expressions
-dynamically. Informational type checking will be provided (except for aggregated
-API server types), but will be surfaced only as warnings. See "Alternatives
-Considered" section for details of all the alternatives we reviewed when
-selecting this approach.
+Due to these complications, Type checking will be optional which can be activated by an expression-level
+`TypeCheckingMode` field. Possible modes are:
 
-Type checking is still performed for all expressions where a GVK can be matched
-to type check against, resulting in warnings, e.g.:
+- `Never` Do not type check the referred object or the param any time. 
+  Treat `object`, `oldObject`, and `params` as dynamic-typed unstructured objects. 
+  This behavior matches that of Kubernetes 1.26
+- `Evaluation` Type check the referred `object`, `oldObject`, and `params` when the expression is being compiled and evaluated after the policy is created. 
+  If the type check fails, mark the policy as misconfigured. Do not type check when the policy is being created or updated.
+- `Always` Type check the referred `object`, `oldObject`, and `params` immediately when the policy is being created or updated. 
+  If the type check fails, including the situation where the schema fails to resolve, the API server will reject the operation on the policy object. 
+  The same type check will happen during compilation and evaluation.
```

Comment:
The difference of these two would be whether we fail at `ValidatingAdmissionPolicy` creation. One special case I can think of would be when the user creates the policy first then the CRD. If there is a re-try loop, making it fail fast seems better, otherwise `Evaluation` would make more sense.
What do you thin?

Reply:
We should implement the modes in the order of `Never`, `Evaluation`, and `Always` while gather user input between releases.

Reply:
SGTM. I like the idea of introducing `Never` and `Evaluation` first and seeing how it goes before investing in type checking on ValidatingAdmissionPolicy create/update.